File: explorer/client/src/util/diffexpdu/postingslist.ts
/* eslint-disable no-bitwise -- relies on bitwise ops */

/*
The format is documented in `dev_docs/diffexpdu.md`. It is implemented in both Python
and Typescript.  Implementations must maintain compatibility.
*/

import { PduBuffer } from "./pdubuffer";
import {
  compressBuffer,
  decompressBuffer,
  nonzeroBits,
  pairwise,
  groupby,
  concat,
} from "./utils";

interface ListPartition {
  listId: number;
  key: number;
  startIdx: number;
  endIdx: number;
  arr: Uint32Array;
}

const MAGIC_NUMBER = 0xce;

enum BlockType {
  BitArray = 0,
  Uint16Array = 1,
  Uint16ArrayInverted = 2,
}

interface BlockDescription {
  blockType: BlockType;
  listIdMask: number;
  nElem: number;
  key: number;
  nBytes: number;
}

export function packPostingsLists(
  pdu: PduBuffer,
  postingsLists: Array<Uint32Array>,
  sorted = false,
): void {
  if (postingsLists.length === 0) {
    throw new Error("must specify one or more postings lists");
  }
  if (!sorted) {
    throw new Error("all postings lists must be sorted");
  }

  const allPartitions = partitionLists(postingsLists);
  allPartitions.sort((a, b) => a.key - b.key);

  const encodedBlocks = [];
  for (const partition of allPartitions) {
    encodedBlocks.push(encodePartition(partition));
  }

  // encode list header
  const blockCount = encodedBlocks.length;
  packPostingsListsHeader(pdu, 1, blockCount);

  // encode each block description
  for (const [header] of encodedBlocks) {
    packBlockDescription(pdu, header);
  }

  // encode each block
  for (const [, block] of encodedBlocks) {
    packBlock(pdu, block);
  }
}

function packPostingsListsHeader(
  pdu: PduBuffer,
  nLists: number,
  nBlocks: number,
): void {
  pdu.ensureFreeCapacity(4);
  pdu.addUint8(MAGIC_NUMBER);
  pdu.addUint8(nLists - 1);
  pdu.addUint16(nBlocks - 1);
}

function packBlockDescription(
  pdu: PduBuffer,
  description: BlockDescription,
): void {
  pdu.ensureFreeCapacity(8);
  pdu.addUint8(description.blockType);
  pdu.addUint8(description.listIdMask);
  pdu.addUint16(description.nElem - 1);
  pdu.addUint16(description.key);
  pdu.addUint16(description.nBytes - 1);
}

function packBlock(pdu: PduBuffer, block: Uint8Array): void {
  pdu.ensureFreeCapacity(block.byteLength);
  pdu.add(block);
}

/**
 * Given a partition, choose a block type which is likely near-optimal for
 * encoding.  Heuristics live here.
 *
 * A description of the heuristics and their rationale is documented
 * in the format description (`dev_docs/diffexpdu.md`)
 *
 * @param partition
 * @returns the type of block
 */
function chooseBlockType(partition: ListPartition): BlockType {
  const nElem = partition.endIdx - partition.startIdx;
  const interval =
    (partition.arr[partition.endIdx - 1] & 0xffff) +
    1 -
    (partition.arr[partition.startIdx] & 0xffff);
  const density = nElem / interval;

  const minBitArrayThreshold = 2 ** 11;
  const maxBitArrayThreshold = 2 ** 16 - minBitArrayThreshold;
  const minDensity = 0.125;
  const maxDensity = 0.875;
  if (
    minBitArrayThreshold < nElem &&
    nElem < maxBitArrayThreshold &&
    minDensity < density &&
    density < maxDensity
  )
    return BlockType.BitArray;

  return BlockType.Uint16Array;
}

function encodePartition(
  partition: ListPartition,
): [BlockDescription, Uint8Array] {
  const arr = partition.arr.subarray(partition.startIdx, partition.endIdx);
  const blockType = chooseBlockType(partition);
  if (blockType === BlockType.BitArray) {
    return encodeBitarrayBlock(partition, arr);
  }
  return encodeUint16Block(partition, arr);
}

function encodeBitarrayBlock(
  partition: ListPartition,
  list: Uint32Array,
): [BlockDescription, Uint8Array] {
  const buf = new PduBuffer(2 ** 13);
  buf.advance(2 ** 13);
  const buf32 = buf.asUint32Array();
  for (let i = 0; i < list.length; i += 1) {
    const val = list[i] & 0xffff;
    buf32[(val / 32) >> 0] |= 1 << (val & 0x1f);
  }
  const compressedBuf = compressBuffer(buf.asUint8Array());
  const header = {
    blockType: BlockType.BitArray,
    listIdMask: 1 << partition.listId,
    nElem: list.length,
    key: list[0] >> 16,
    nBytes: compressedBuf.length,
  };
  return [header, compressedBuf];
}

function encodeUint16Block(
  partition: ListPartition,
  list: Uint32Array,
): [BlockDescription, Uint8Array] {
  const nElem = list.length;
  const key = list[0] >> 16;
  let blockType: BlockType;
  let buf: Uint8Array;

  const nElemInverted = intervalInvertedLength(list);

  if (nElemInverted + 2 >= nElem) {
    // Encode as-is.
    blockType = BlockType.Uint16Array;
    buf = byteShuffle16(delta16(list));
  } else {
    // Encode the inverted values as it will be shorter.
    // inverted array is prefaced by [intervalStart, intervalEnd]
    blockType = BlockType.Uint16ArrayInverted;
    buf = new Uint8Array(2 * nElemInverted + 4);
    const dv = new DataView(buf.buffer);
    dv.setUint16(0, list[0] & 0xffff, true); // interval start
    dv.setUint16(2, list[list.length - 1] & 0xffff, true); // interval end INCLUSIVE
    byteShuffle16(delta16(intervalInvert(list)), buf.subarray(4));
  }

  const compressedBuf = compressBuffer(buf);
  const header = {
    blockType,
    listIdMask: 1 << partition.listId,
    nElem,
    key,
    nBytes: compressedBuf.length,
  };
  return [header, compressedBuf];
}

export function unpackPostingsLists(buf: Uint8Array): Array<Uint32Array> {
  const dv = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  const [startOffset, , , blockDescriptions] = unpackPostingsListsHeader(dv);

  let offset = startOffset;
  const subLists: [number, Uint32Array][] = [];
  for (const blockDescription of blockDescriptions) {
    const subList = unpackBlock(blockDescription, buf.subarray(offset));
    subLists.push(...subList);
    offset += blockDescription.nBytes;
  }

  // now group by listId and concatenate
  subLists.sort((a, b) => a[0] - b[0]);
  const concatLists = [];
  for (const listGroup of groupby(subLists, (sl) => sl[0])) {
    concatLists.push(
      concat(
        Uint32Array,
        listGroup.map((lg) => lg[1]),
      ),
    );
  }
  return concatLists;
}

function unpackPostingsListsHeader(
  dv: DataView,
): [number, number, number, Array<BlockDescription>] {
  const magic = dv.getUint8(0);
  if (magic !== MAGIC_NUMBER)
    throw new Error("Malformed postings list header - incorrect magic number");
  const nLists = dv.getUint8(1) + 1;
  const nBlocks = dv.getUint16(2, true) + 1;
  let offset = 4;

  const blockDescriptions = [];
  for (let i = 0; i < nBlocks; i += 1) {
    const blockType = dv.getUint8(offset + 0);
    const listIdMask = dv.getUint8(offset + 1);
    const nElem = dv.getUint16(offset + 2, true) + 1;
    const key = dv.getUint16(offset + 4, true);
    const nBytes = dv.getUint16(offset + 6, true) + 1;
    blockDescriptions.push({
      blockType,
      listIdMask,
      nElem,
      key,
      nBytes,
    });
    offset += 8;
  }
  return [offset, nLists, nBlocks, blockDescriptions];
}

function unpackBlock(
  blockDescription: BlockDescription,
  buf: Uint8Array,
): Array<[number, Uint32Array]> {
  const { blockType } = blockDescription;
  const decoded: [number, Uint32Array][] = []; // array of [listId, list]
  const listIds = nonzeroBits(blockDescription.listIdMask);
  // Current spec only supports one list per block.
  if (listIds.length !== 1) throw new Error("Unexpected multi-list block.");

  buf = decompressBuffer(buf);

  switch (blockType) {
    case BlockType.BitArray: {
      decoded.push(decodeBitarrayBlock(blockDescription, buf, listIds));
      break;
    }
    case BlockType.Uint16Array: {
      decoded.push(decodeUint16Block(blockDescription, buf, listIds));
      break;
    }
    case BlockType.Uint16ArrayInverted: {
      decoded.push(decodeUint16InvertedBlock(blockDescription, buf, listIds));
      break;
    }
    default:
      throw new Error("Unknown block type");
  }

  // Add key to each block result
  const key = blockDescription.key << 16;
  const { nElem } = blockDescription;
  for (const [, list] of decoded) {
    if (list.length !== nElem)
      throw new Error("Block corrupted - nElem mismatch");
    for (let i = 0; i < nElem; i += 1) {
      list[i] += key;
    }
  }

  return decoded;
}

function decodeBitarrayBlock(
  blockDescription: BlockDescription,
  buf: Uint8Array,
  listIds: number[],
): [number, Uint32Array] {
  const { nElem } = blockDescription;
  const buf32 = new Uint32Array(buf.buffer, buf.byteOffset);
  const list = new Uint32Array(blockDescription.nElem);
  for (let i = 0, elem = 0; i < 2 ** 16 && elem < nElem; i += 1) {
    if (buf32[(i / 32) >> 0] & (1 << (i & 0x1f))) {
      list[elem] = i;
      elem += 1;
    }
  }
  return [listIds[0], list];
}

function decodeUint16Block(
  _: BlockDescription,
  buf: Uint8Array,
  listIds: number[],
): [number, Uint32Array] {
  return [listIds[0], unDelta16(unByteShuffle16(buf))];
}

function decodeUint16InvertedBlock(
  _: BlockDescription,
  buf: Uint8Array,
  listIds: number[],
): [number, Uint32Array] {
  const dv = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  const interval: [number, number] = [
    dv.getUint16(0, true),
    dv.getUint16(2, true) + 1,
  ];
  const list = intervalInvert(
    unDelta16(unByteShuffle16(buf.subarray(4))),
    interval,
  );
  return [listIds[0], list];
}

function partitionLists(plists: Array<Uint32Array>): Array<ListPartition> {
  const partitions = [];
  for (let listId = 0; listId < plists.length; listId += 1) {
    const plist = plists[listId];
    const boundaries = findPartitionBoundaries(plist);
    for (const [startIdx, endIdx] of pairwise(boundaries)) {
      partitions.push({
        listId,
        key: plist[startIdx] >> 16,
        startIdx,
        endIdx,
        arr: plist,
      });
    }
  }
  return partitions;
}

function findPartitionBoundaries(arr: Uint32Array): Array<number> {
  const partitions: number[] = [];
  if (arr.length === 0) return partitions;

  let currentPartition = -1;
  for (let idx = 0; idx < arr.length; idx += 1) {
    const val = arr[idx];
    const partition = val >> 16;
    if (partition !== currentPartition) {
      partitions.push(idx);
      currentPartition = val >> 16;
    }
  }
  partitions.push(arr.length); // terminal
  return partitions;
}

/**
 * Compute delta between each element, for lowest 16 bits ONLY.
 *
 * @param src - Uint32Array, sorted, with common top 16 bits.
 * @returns Uint16Array, with difference by position
 */
function delta16(src: Uint32Array): Uint16Array {
  const nElem = src.length;
  const dst = new Uint16Array(nElem);
  let prev = src[0] & 0xffff0000;
  let i: number;

  // unroll loop by blocks of eight
  for (i = 0; i < (nElem & ~7); i += 8) {
    dst[i + 7] = src[i + 7] - src[i + 6];
    dst[i + 6] = src[i + 6] - src[i + 5];
    dst[i + 5] = src[i + 5] - src[i + 4];
    dst[i + 4] = src[i + 4] - src[i + 3];
    dst[i + 3] = src[i + 3] - src[i + 2];
    dst[i + 2] = src[i + 2] - src[i + 1];
    dst[i + 1] = src[i + 1] - src[i + 0];
    dst[i + 0] = src[i + 0] - prev;
    prev = src[i + 7];
  }
  for (; i < nElem; i += 1) {
    dst[i] = src[i] - prev;
    prev = src[i];
  }
  return dst;
}

function unDelta16(src: Uint16Array): Uint32Array {
  const nElem = src.length;
  const dst = new Uint32Array(nElem);
  let prev = 0;
  for (let i = 0; i < nElem; i += 1) {
    dst[i] = prev + src[i];
    prev = dst[i];
  }
  return dst;
}

function byteShuffle16(src: Uint16Array, dst?: Uint8Array): Uint8Array {
  const nElem = src.length;
  const srcU8 = new Uint8Array(src.buffer);
  if (dst === undefined) dst = new Uint8Array(2 * nElem);

  // unroll loop by blocks of four
  let i = 0;
  for (; i < (nElem & ~3); i += 4) {
    dst[i + 0] = srcU8[(i + 0) * 2];
    dst[i + 1] = srcU8[(i + 1) * 2];
    dst[i + 2] = srcU8[(i + 2) * 2];
    dst[i + 3] = srcU8[(i + 3) * 2];

    dst[nElem + (i + 0)] = srcU8[(i + 0) * 2 + 1];
    dst[nElem + (i + 1)] = srcU8[(i + 1) * 2 + 1];
    dst[nElem + (i + 2)] = srcU8[(i + 2) * 2 + 1];
    dst[nElem + (i + 3)] = srcU8[(i + 3) * 2 + 1];
  }
  for (; i < nElem; i += 1) {
    dst[i] = srcU8[i * 2];
    dst[nElem + i] = srcU8[i * 2 + 1];
  }

  return dst;
}

function unByteShuffle16(src: Uint8Array, dst?: Uint8Array): Uint16Array {
  const typeSize = 2; // Uint16
  const nElem = src.length / typeSize;
  if (dst === undefined) dst = new Uint8Array(src.length);

  for (let i = 0; i < nElem; i += 1) {
    for (let j = 0; j < typeSize; j += 1) {
      dst[i * typeSize + j] = src[j * nElem + i];
    }
  }
  return new Uint16Array(dst.buffer);
}

/**
 * For the given increasing monotone `arr`, return the length of
 * the array required to store the array if it was inverted across
 * the given interval.
 *
 * Examples:
 *    [0, 4] -> 3
 *    [0, 1, 3] -> 1
 *
 * @param arr - Uint32Array, containing an increasing monotone.
 * @param interval - the interval across which to invert the values. If not specified, defaults to the range of values in `arr`.
 */
function intervalInvertedLength(
  arr: Uint32Array,
  interval?: [number, number],
): number {
  if (interval === undefined) {
    interval = [arr[0], arr[arr.length - 1] + 1];
  }
  return interval[1] - interval[0] - arr.length;
}

/**
 * Invert the array values over the given interval. Array must be sorted monotone.
 * If interval not specified, will use [arr[0], arr[-1]+1]
 * @param src - source array
 * @param interval - [start, end)
 * @returns inverted array
 */
function intervalInvert(
  src: Uint32Array,
  interval?: [number, number],
): Uint32Array {
  if (interval === undefined) {
    interval = [src[0], src[src.length - 1] + 1];
  }
  const dst = new Uint32Array(intervalInvertedLength(src, interval));
  for (let i = interval[0], srcIdx = 0, dstIdx = 0; i < interval[1]; i += 1) {
    if (srcIdx < src.length && src[srcIdx] === i) {
      srcIdx += 1;
    } else {
      dst[dstIdx] = i;
      dstIdx += 1;
    }
  }
  return dst;
}

/* eslint-enable no-bitwise -- relies on bitwise ops */

File: explorer/server/common/diffexpdu/postingslist.py
import itertools
import struct
from dataclasses import dataclass
from enum import IntEnum
from typing import ClassVar, List, Tuple, Type, TypeVar, Union

import bitarray  # https://github.com/ilanschnell/bitarray
import bitarray.util
import numba as nb
import numpy as np

from .utils import _blockCompress, _blockDecompress, _nonzero_bits, pairwise

"""
The format is documented in `dev_docs/diffexpdu.md`. It is implemented in both Python
and Typescript.  Implementations must maintain compatibility.
"""

PostingsList = Union[np.ndarray, List[int]]
DisjointPostingsLists = Tuple[PostingsList]
ListId = int

T = TypeVar("T")
MAGIC_NUMBER = 0xCE

headerPacker = struct.Struct("<BBH")
blockDescriptionPacker = struct.Struct("<BBHHH")
uint16 = struct.Struct("<H")


@dataclass
class _BlockDescription:
    class BlockType(IntEnum):
        BitArray = 0
        Uint16Array = 1
        Uint16ArrayInverted = 2

    block_type: BlockType
    list_id_mask: int  # 8 bit bitmask
    n_elem: int  # n elements in block
    key: int  # top two bytes (16 MSB)
    n_bytes: int  # encoded length in bytes, excluding header

    packed_length: ClassVar[int] = blockDescriptionPacker.size

    @classmethod
    def from_bytes(cls: Type[T], buf: Union[bytes, bytearray]) -> T:
        assert len(buf) >= blockDescriptionPacker.size
        block_type, list_id_mask, n_elem, key, n_bytes = blockDescriptionPacker.unpack_from(buf)
        return _BlockDescription(
            block_type=block_type, list_id_mask=list_id_mask, n_elem=n_elem + 1, key=key, n_bytes=n_bytes + 1
        )

    def to_bytes(self) -> bytes:
        return blockDescriptionPacker.pack(
            self.block_type, self.list_id_mask, self.n_elem - 1, self.key, self.n_bytes - 1
        )

    def list_ids(self) -> List[int]:
        """
        Return list IDs as list of ints
        """
        return _nonzero_bits(self.list_id_mask)


@dataclass
class _ListPartition:
    list_id: int  # 0-7, list identifier
    key: int
    start_idx: int
    end_idx: int  # exclusive range, eg, [start_idx, end_idx)
    arr: np.ndarray  # the full postings list


def packed_length(buf: Union[bytes, bytearray, memoryview], offset=0) -> int:
    """
    Return the length of the deflated posting list encoded in buf.

    Returns
    -------
    int
        Length of deflated posting list in bytes.
    """
    buf = memoryview(buf)
    if offset > 0:
        buf = buf[offset:]

    _, block_descriptions = _decode_header(buf)
    n_blocks = len(block_descriptions)
    return (
        headerPacker.size
        + n_blocks * _BlockDescription.packed_length
        + sum([block.n_bytes for block in block_descriptions])
    )


def deflate_postings_lists(disjoint_postings_lists: DisjointPostingsLists, sorted: bool = False) -> bytearray:
    """
    Efficiently pack one or more (max 8) monotonically increasing lists of unsigned integers into a byte stream.
    Lists must be disjoint (the current implementation may not fail if lists are not unique, but future
    changes may enforce/require this)
    """
    assert 0 < len(disjoint_postings_lists) <= 8

    def normalize_plist(pl):
        if isinstance(pl, list):
            pl = np.array(pl, dtype=np.uint32)
        if pl.dtype != np.uint32:
            pl = pl.astype(np.uint32)
        if not sorted:
            pl.sort()
        return pl

    disjoint_lists = tuple((normalize_plist(plist) for plist in disjoint_postings_lists))
    all_partitions = partition_lists(disjoint_lists)
    all_partitions.sort(key=lambda partition: partition.key)

    encoded_blocks = []
    for _, parts in itertools.groupby(all_partitions, key=lambda partition: partition.key):
        # encode each list as individual blocks.
        parts_tuple = tuple(parts)
        assert len(parts_tuple) <= 8
        encoded_blocks.extend(_encode_partition(part) for part in parts_tuple)

    block_count = len(encoded_blocks)
    encoded_postings = bytearray(4)
    headerPacker.pack_into(encoded_postings, 0, MAGIC_NUMBER, 0, block_count - 1)

    # encode block descriptions
    for hdr, values in encoded_blocks:
        assert hdr.n_bytes == len(values)
        encoded_postings.extend(hdr.to_bytes())

    assert len(encoded_postings) == headerPacker.size + blockDescriptionPacker.size * block_count

    # encode blocks
    for _, values in encoded_blocks:
        encoded_postings.extend(values)

    return encoded_postings


def _choose_block_type(part: _ListPartition) -> _BlockDescription.BlockType:
    """
    Given a partition, choose a block type which is likely near-optimal for
    encoding.  Heuristics live here.

    A description of the heuristics and their rationale is documented
    in the format description (`dev_docs/diffexpdu.md`)
    """
    n_elem = part.end_idx - part.start_idx
    assert 0 < n_elem <= 2**16
    assert (part.arr[part.start_idx] >> 16) == (part.arr[part.end_idx - 1] >> 16)

    interval = (part.arr[part.end_idx - 1] & 0xFFFF) + 1 - (part.arr[part.start_idx] & 0xFFFF)
    density = n_elem / interval

    minBitArrayThreshold = 2**11
    maxBitArrayThreshold = 2**16 - minBitArrayThreshold
    if (minBitArrayThreshold < n_elem < maxBitArrayThreshold) and (0.125 < density < 0.875):
        return _BlockDescription.BlockType.BitArray

    return _BlockDescription.BlockType.Uint16Array


def _encode_partition(part: _ListPartition) -> Tuple[_BlockDescription, bytearray]:
    """for a slice of the posting list, create a serialized block"""
    assert part.arr.dtype.kind == "u"
    arr = part.arr[part.start_idx : part.end_idx].astype(np.uint16)
    block_type = _choose_block_type(part)
    if block_type == _BlockDescription.BlockType.Uint16Array:
        return _encode_uint16_block(part, arr)
    if block_type == _BlockDescription.BlockType.BitArray:
        return _encode_bitarray_block(part, arr)

    raise Exception("Unknown block type prediction")


def _encode_bitarray_block(part: _ListPartition, arr: np.ndarray) -> Tuple[_BlockDescription, bytearray]:
    """
    Encode as bitarray, compressed with zlib.
    """
    bool_arr = np.zeros((2**16,), dtype=np.bool_)
    bool_arr[arr] = 1
    ba = bitarray.bitarray(endian="little")
    ba.pack(bool_arr.tobytes())

    buf = ba.tobytes()
    assert len(buf) == 2**13
    buf = _blockCompress(buf)
    key = part.arr[part.start_idx] >> 16
    header = _BlockDescription(
        block_type=_BlockDescription.BlockType.BitArray,
        list_id_mask=(1 << part.list_id),
        n_elem=len(arr),
        key=key,
        n_bytes=len(buf),
    )
    return (header, buf)


def _encode_uint16_block(part: _ListPartition, arr: np.ndarray) -> Tuple[_BlockDescription, bytearray]:
    """
    Encode lower two bytes only. Applies steps:
        * optional inversion
        * delta codes
        * byte shuffle
        * zlib.compress
    """
    key = part.arr[part.start_idx] >> 16
    n_elem = len(arr)

    if _interval_inverted_length(arr) + 2 >= n_elem:
        block_type = _BlockDescription.BlockType.Uint16Array
        arr = _byteshuffle(_delta(arr))
        buf = arr.tobytes()
        assert len(buf) == 2 * n_elem
    else:
        # invert the array if it would be shorter to encode the
        # range over which it is inverted and the elements not
        # in the original
        block_type = _BlockDescription.BlockType.Uint16ArrayInverted
        buf = bytearray(4)
        uint16.pack_into(buf, 0, int(arr[0]))  # range start
        uint16.pack_into(buf, 2, int(arr[-1]))  # range end INCLUSIVE
        arr = _byteshuffle(_delta(_interval_invert(arr)))
        buf.extend(arr.tobytes())
        assert len(buf) == 4 + len(arr)

    buf = _blockCompress(buf)
    header = _BlockDescription(
        block_type=block_type, list_id_mask=(1 << part.list_id), n_elem=n_elem, key=key, n_bytes=len(buf)
    )
    return (header, buf)


def inflate_postings_lists(buf: Union[bytes, bytearray, memoryview], offset=0) -> List[np.ndarray]:
    """
    Decode a serialized postings list, returning the original list as a ``numpy.ndarray``.
    """
    buf = memoryview(buf)
    n_lists, block_descriptions = _decode_header(buf[offset:])
    n_blocks = len(block_descriptions)
    offset += headerPacker.size + n_blocks * _BlockDescription.packed_length

    # decode each block
    sub_lists = []
    for block_description in block_descriptions:
        sub_lists.extend(_decode_block(block_description, buf[offset : offset + block_description.n_bytes]))
        offset += block_description.n_bytes

    sub_lists.sort(key=lambda sl: sl[0])  # sort by list id
    concat_lists = [
        np.concatenate([g[1] for g in grp]) for _, grp in itertools.groupby(sub_lists, key=lambda sl: sl[0])
    ]
    return concat_lists


def _decode_header(buf: memoryview) -> Tuple[int, List[_BlockDescription]]:
    magic, n_lists, n_blocks = headerPacker.unpack_from(buf)
    offset = headerPacker.size
    n_blocks += 1
    n_lists += 1
    assert magic == MAGIC_NUMBER
    assert n_blocks > 0
    assert n_lists > 0

    # decode block descriptions
    block_descriptions = [
        _BlockDescription.from_bytes(buf[b:])
        for b in range(
            offset,
            offset + n_blocks * _BlockDescription.packed_length,
            _BlockDescription.packed_length,
        )
    ]
    return n_lists, block_descriptions


def _decode_block(desc: _BlockDescription, buf: memoryview) -> List[Tuple[ListId, np.ndarray]]:
    block_type = desc.block_type
    n_elem = desc.n_elem
    buf = _blockDecompress(buf)
    list_ids = desc.list_ids()

    decoded = []  # list of (ListId, ndarray)
    if block_type == _BlockDescription.BlockType.BitArray:
        decoded.append(_decode_bitarray_block(buf, list_ids))

    elif block_type == _BlockDescription.BlockType.Uint16Array:
        decoded.append(_decode_uint16_block(buf, list_ids))

    elif block_type == _BlockDescription.BlockType.Uint16ArrayInverted:
        decoded.append(_decode_uint16_inverted_block(buf, list_ids))

    else:
        raise AssertionError("Unknown block type")

    assert len(decoded) > 0
    assert sum(len(d[1]) for d in decoded) == n_elem
    decoded = [(list_id, arr.astype(np.uint32) + (desc.key << 16)) for list_id, arr in decoded]
    return decoded


def _decode_bitarray_block(buf, list_ids):
    assert len(buf) == 8192, f"Unexpected bitarray length {len(buf)}"
    assert len(list_ids) == 1
    ba = bitarray.bitarray(endian="little")
    ba.frombytes(buf)
    arr = np.frombuffer(ba.unpack(), dtype=np.bool_).nonzero()[0]
    return (list_ids[0], arr)


def _decode_uint16_block(buf, list_ids):
    assert len(list_ids) == 1
    arr = _un_delta(_un_byteshuffle(buf, dtype=np.uint16))
    return (list_ids[0], arr)


def _decode_uint16_inverted_block(buf, list_ids):
    assert len(list_ids) == 1
    interval = (
        uint16.unpack_from(buf, 0)[0],
        uint16.unpack_from(buf, 2)[0] + 1,
    )
    arr = _interval_invert(_un_delta(_un_byteshuffle(buf[4:], dtype=np.uint16)), interval=interval)
    return (list_ids[0], arr)


def _byteshuffle(arr: np.ndarray) -> np.ndarray:
    """classic byteshuffle to improve compressibility.  Always in-place."""
    return arr.view(np.uint8).reshape((arr.itemsize, len(arr)), order="F").flatten()


def _un_byteshuffle(buf: Union[bytes, bytearray, memoryview], dtype) -> np.ndarray:
    """invert byteshuffle, in-place."""
    dtype = np.dtype(dtype)
    arr = np.frombuffer(buf, dtype=np.uint8)
    return arr.reshape(len(buf) // dtype.itemsize, dtype.itemsize, order="F").flatten().view(dtype)


def _delta(arr: np.ndarray) -> np.ndarray:
    return np.diff(arr, prepend=arr.dtype.type(0))


def _un_delta(arr: np.ndarray) -> np.ndarray:
    return np.cumsum(arr)


@nb.jit
def _interval_invert_inner(src: np.ndarray, dst: np.ndarray, interval: Tuple[int, int]) -> np.ndarray:
    src_idx = 0
    dst_idx = 0
    for i in range(interval[0], interval[1]):
        if src_idx < len(src) and src[src_idx] == i:
            src_idx += 1
            continue
        dst[dst_idx] = i
        dst_idx += 1
    return dst


def _interval_invert(arr: np.ndarray, interval=None) -> np.ndarray:
    """Invert over an interval."""
    assert arr.dtype.kind == "u"
    assert interval is not None or len(arr) > 0
    if interval is None:
        interval = (arr[0], arr[-1] + 1)
    assert len(arr) == 0 or (arr[0] >= interval[0] and arr[0] < interval[1])
    assert len(arr) == 0 or (arr[-1] >= interval[0] and arr[-1] < interval[1])
    return _interval_invert_inner(
        arr, np.empty_like(arr, shape=(_interval_inverted_length(arr, interval=interval),)), interval
    )


def _interval_inverted_length(arr: np.ndarray, interval=None) -> int:
    """
    For the given array, return the length if it was inverted across
    the interval of actual values.

    Assert: arr is an increasing monotone.
    """
    if interval is None:
        interval = (arr[0], arr[-1] + 1)
    return interval[1] - interval[0] - len(arr)


@nb.jit
def _find_partition_boundaries(arr: np.ndarray) -> List[int]:
    """
    Given an array of monotonically increasing unsigned ints, return the
    indices where the most significant 16 bits changes.

    Currently a simple scan.
    """
    partitions = []
    if len(arr) == 0:
        return partitions

    current_partition = -1
    for idx in range(len(arr)):
        val = arr[idx]
        partition = val >> 16
        if partition == current_partition:
            continue

        partitions.append(idx)
        current_partition = val >> 16

    partitions.append(len(arr))  # terminal
    return partitions


def partition_lists(plists: DisjointPostingsLists) -> List[_ListPartition]:
    assert 0 < len(plists) <= 8
    partitions = []
    for list_id in range(len(plists)):
        plist = plists[list_id]
        if isinstance(plist, list):
            plist = np.array(plist, dtype=np.uint32)
        boundaries = _find_partition_boundaries(plist)
        partitions += [
            _ListPartition(
                list_id=list_id, key=(plist[start_idx] >> 16), start_idx=start_idx, end_idx=end_idx, arr=plist
            )
            for start_idx, end_idx in pairwise(boundaries)
        ]
    return partitions
